### PR TITLE
Database

### PR DESCRIPTION
**Pull request by JustinHsie**
Hey Christian!

I made the database changes, sorry for the delay! Ran into a lot of problems haha

Changes:
- Connect to PostgreSQL database
- Push database to Heroku and connect app

It turned into a bit of a headache because I realized I'd been storing things in the mockDb as objects this whole time, which doesn't translate to a SQL database. 
I didn't want to redo my entire frontend, so I had to make multiple db tables to represent those objects and make separate SQL queries in order to 'compile' the data into what the frontend was expecting.

Pushing the db to Heroku was mostly fine, just took a day to figure out. 
And the working app is on the production site now (not dev) bc I wanted the main app to have the database.

Again, please don't spend too much time looking at everything haha, I know it probably shouldn't have been done this way.
Thanks so much!!

### CODE REVIEW

**Comment by Exegetech** [comment]
> Hey Christian!
> 
> I made the database changes, sorry for the delay! Ran into a lot of problems haha
> 
> Changes:
> 
>     * Connect to PostgreSQL database
> 
>     * Push database to Heroku and connect app
> 
> 
> It turned into a bit of a headache because I realized I'd been storing things in the mockDb as objects this whole time, which doesn't translate to a SQL database.

This is the entire point of the exercise of why I told you to do MockDb. When you change the underlying implementation, ideally you should make the abstraction in such a way that it is quite easy to pull the MockDB out and replace it with real database with a few changes.

> I didn't want to redo my entire frontend, so I had to make multiple db tables to represent those objects and make separate SQL queries in order to 'compile' the data into what the frontend was expecting.
> 

See above. If you have your abstractions are correct it shouldn't require you to change the frontend that much. This means you database implementation is leaking all the way through frontend. 

> Pushing the db to Heroku was mostly fine, just took a day to figure out.
> And the working app is on the production site now (not dev) bc I wanted the main app to have the database.
> 

I don't see a migration script. How are you planning to do migration? Also, say a new developer onboard and work on this, how would he/she prepare the database locally? Migration script is needed for this.

> Again, please don't spend too much time looking at everything haha, I know it probably shouldn't have been done this way.
> Thanks so much!!

It is already too much to see at this point. I won't nitpick as much, but you are currently writing a raw SQL query. In real world, this is a no no, you'd be vulnerable to SQL injection attacks. You have to sanitize your SQL. Use library like knex or an ORM library to mitigate this. It is your choice whether in this toy project you want to use knex or not, but just beware in real world this is not acceptable.

**Comment by JustinHsie** [comment]
> This is the entire point of the exercise of why I told you to do MockDb. When you change the underlying implementation, ideally you should make the abstraction in such a way that it is quite easy to pull the MockDB out and replace it with real database with a few changes.
> 
Yeah I remember you saying that...shoot, was I supposed to make my own MockDb? 
I used the example you made me, but at the time I didn't realize it's not a SQL implementation. (eg I shouldn't need a method "add place to list" because each place would have a foreign key that references the lists table). 
Is it possible to use the MockDb in a SQL way?

> I don't see a migration script. How are you planning to do migration? Also, say a new developer onboard and work on this, how would he/she prepare the database locally? Migration script is needed for this.
> 
Oh, I'd never heard of a migration script before. I just looked it up briefly, so from what I read it's basically a build script or additional scripts if I want to change the db schema.

Am I supposed to run the migration script from my backend to build the db in Heroku? Or is it just a separate script for reference?
Because what I did was push up a pre-made db for Heroku to host.


> It is already too much to see at this point. I won't nitpick as much, but you are currently writing a raw SQL query. In real world, this is a no no, you'd be vulnerable to SQL injection attacks. You have to sanitize your SQL. Use library like knex or an ORM library to mitigate this. It is your choice whether in this toy project you want to use knex or not, but just beware in real world this is not acceptable.
>
Maybe I misread something, but I'm using node-postgres and I think the library sanitizes parameterized queries written in that syntax. But that's good to know, thanks. 

Sorry, I should've asked you what to do first before doing all this. For some reason I thought you'd be against the abstraction of ORMs lol so I chose to write SQL. I also thought it'd be valuable for employers to know that I can write SQL.

**Comment by Exegetech** [comment]
> > This is the entire point of the exercise of why I told you to do MockDb. When you change the underlying implementation, ideally you should make the abstraction in such a way that it is quite easy to pull the MockDB out and replace it with real database with a few changes.
> 
> Yeah I remember you saying that...shoot, was I supposed to make my own MockDb?
> I used the example you made me, but at the time I didn't realize it's not a SQL implementation. (eg I shouldn't need a method "add place to list" because each place would have a foreign key that references the lists table).
> Is it possible to use the MockDb in a SQL way?

It is possible, that's how ORM is made. So technically you are creating your own ORM. But I guess no need to worry about it for now, just do whatever needs to be done.

> 
> > I don't see a migration script. How are you planning to do migration? Also, say a new developer onboard and work on this, how would he/she prepare the database locally? Migration script is needed for this.
> 
> Oh, I'd never heard of a migration script before. I just looked it up briefly, so from what I read it's basically a build script or additional scripts if I want to change the db schema.
> 
> Am I supposed to run the migration script from my backend to build the db in Heroku? Or is it just a separate script for reference?
> Because what I did was push up a pre-made db for Heroku to host.
> 

It is a build script if you or another developer wants to create or change the schema. How else you are going to ask the developer or you yourself to recreate the db from scratch?

I am not aware what Heroku's way of doing db, but I imagine yes, you still need migration script, otherwise how are you going to change the schema when the database is up and running with user's data?

Migration script only makes sense in a schema-full database, which in this case SQL. In NoSQL database, schema sometimes not necessary (I say sometimes because this isn't true for all NoSQL databases).

> > It is already too much to see at this point. I won't nitpick as much, but you are currently writing a raw SQL query. In real world, this is a no no, you'd be vulnerable to SQL injection attacks. You have to sanitize your SQL. Use library like knex or an ORM library to mitigate this. It is your choice whether in this toy project you want to use knex or not, but just beware in real world this is not acceptable.
> 
> Maybe I misread something, but I'm using node-postgres and I think the library sanitizes parameterized queries written in that syntax. But that's good to know, thanks.
> 

Yeah I wasn't aware of that, I just read the docs and it seems so, so this is fine.

> Sorry, I should've asked you what to do first before doing all this. For some reason I thought you'd be against the abstraction of ORMs lol so I chose to write SQL. I also thought it'd be valuable for employers to know that I can write SQL.

I myself don't like ORM, and this is your project so you can do whatever you want. Yes you need to know how to write SQL. I didn't see any join statements on the SQL btw.

**Review by Exegetech — APPROVED** [review]

**Comment by JustinHsie** [comment]
> It is a build script if you or another developer wants to create or change the schema. How else you are going to ask the developer or you yourself to recreate the db from scratch?
> 
> I am not aware what Heroku's way of doing db, but I imagine yes, you still need migration script, otherwise how are you going to change the schema when the database is up and running with user's data?
> 
> Migration script only makes sense in a schema-full database, which in this case SQL. In NoSQL database, schema sometimes not necessary (I say sometimes because this isn't true for all NoSQL databases).
> 
I added a migration script, something like this? Then I can call it in the db constructor. 

But if it's not right, can I just include the script for reference but not use it? Because the db is already up and running and everything's connected and working.

And idk if this is related to your question, but Heroku allows me to connect to the remote db locally to make changes. I probably won't be changing the schema anymore though.

> I myself don't like ORM, and this is your project so you can do whatever you want. Yes you need to know how to write SQL. I didn't see any join statements on the SQL btw.
>
The queries were actually pretty simple so I didn't need joins. Plus if I used joins, it would return everything as its own object property, but I needed to return objects within objects

**Comment by Exegetech** [comment]
> > It is a build script if you or another developer wants to create or change the schema. How else you are going to ask the developer or you yourself to recreate the db from scratch?
> > I am not aware what Heroku's way of doing db, but I imagine yes, you still need migration script, otherwise how are you going to change the schema when the database is up and running with user's data?
> > Migration script only makes sense in a schema-full database, which in this case SQL. In NoSQL database, schema sometimes not necessary (I say sometimes because this isn't true for all NoSQL databases).
> 
> I added a migration script, something like this? Then I can call it in the db constructor.
> 
> But if it's not right, can I just include the script for reference but not use it? Because the db is already up and running and everything's connected and working.
> 

It is the right direction but not complete.  Migration refers to the automated process of rolling up and rolling back (if something happens), so you need to have it timestamped and ordered. There is a framework that helps you do this in Knex. But you don't have to implement it if you don't want to. If you call this in the DB constructor then your database will be destroyed everytime you deploy or start the app and thus losing your data.

> And idk if this is related to your question, but Heroku allows me to connect to the remote db locally to make changes. I probably won't be changing the schema anymore though.
> 

A real app most likely undergone schema changes many times. No, you aren't supposed to connect your local db with a production database, I think Heroku only intended this for development purposes. If something happen to your production database then your company is toast. 

> > I myself don't like ORM, and this is your project so you can do whatever you want. Yes you need to know how to write SQL. I didn't see any join statements on the SQL btw.
> 
> The queries were actually pretty simple so I didn't need joins. Plus if I used joins, it would return everything as its own object property, but I needed to return objects within objects

You need joins because it is one trip instead of n trips (network calls are expensive). You can use joins and return whatever shape you want, just do the data transformation in the code on the Db code.

File: server/db/index.js [hunk 1]
```diff
@@ -60,33 +60,40 @@ class Database {
...
-        location.id,
-      ]);
-      const mapCenter = res.rows[0];
+      let res = await this.pool.query(
+        `SELECT places.id AS placeid, places.name AS placename, places.rating, 
+                places.description AS placedescription, locations.name AS locationname, 
+                locations.address, mapcenters.lat, mapcenters.lng, 
+                lists.id AS listid, lists.name AS listname, lists.description AS listdescription 
+         FROM places 
+         LEFT JOIN locations ON places.id = locations.placeid
```
Comment (Exegetech) [inline server/db/index.js]: Location and Place and Map Center has 1 to 1 correlation, and a Place must have a Location and must have a Map Center, am I right? In that case, you can just combine the tables together. Joining isn't exactly free either.

For more info about designing tables, see https://www.guru99.com/database-normalization.html as a guide, depending on the tradeoffs, whether speed, or referential consistency. I think majority of database out there only goes to 3NF.

File: server/db/index.js [hunk 2]
```diff
@@ -170,24 +177,52 @@ class Database {
...
-      res = await this.pool.query('SELECT id FROM places WHERE listid = $1', [
-        listId,
-      ]);
-      const results = res.rows;
-      const places = await Promise.all(
-        results.map(async result => {
-          return await this.getPlace(result.id);
-        })
+      let res = await this.pool.query(
+        `SELECT lists.id AS listid, lists.name AS listname, lists.description AS listdescription,
```
Comment (Exegetech) [inline server/db/index.js:183]: Same as above

File: server/db/index.js (hunk 1 above)
Comment (JustinHsie) [reply server/db/index.js]: Ohh yeahh....wow that was stupid of me...lol

**Comment by JustinHsie** [comment]
Wow thanks so much for the tip on making it a single table...
So changing from the MockDb to Postgres should've been this smooth, but I freaked out when it didn't match exactly and made things waaayy too complicated.

If everything looks ok, I'll merge it!

**Comment by Exegetech** [comment]
Alright looks good, go ahead and merge it.

**Comment by Exegetech** [comment]
What's next? Are all the functions complete?

**Comment by JustinHsie** [comment]
Yeah so at this point the app does everything I set out to do!

In the beginning you talked about adding testing, code coverage, monitoring, logging, and error handling, etc. 
Should I do these next? Are these expected in entry level projects?

And maybe I can add these at a slower pace as I start working on my personal site, a second smaller project, and applying to some places.
If anyone asks, I can say these features are still in development.

**Comment by Exegetech** [comment]
> Yeah so at this point the app does everything I set out to do!
> 
> In the beginning you talked about adding testing, code coverage, monitoring, logging, and error handling, etc.
> Should I do these next? Are these expected in entry level projects?

No, they are not expected in entry level projects.

> 
> And maybe I can add these at a slower pace as I start working on my personal site, a second smaller project, and applying to some places.
> If anyone asks, I can say these features are still in development.

Ok, whatever you want it to be. The app is done then. I'll close the project.

**Pull request merged**